File: BaekjoonOnlineJudge/2512/2512.py
import sys
input = sys.stdin.readline


def dividend(target):
    tot = 0

    for req in require:
        if target >= req:
            tot += req
        else:
            tot += target
            
    return tot


N = int(input())
require = list(map(int, input().split()))
M = int(input())

left = 0
right = max(require)
answer = 0

while left <= right:
    mid = (left + right) // 2

    if dividend(mid) <= M:
        answer = mid
        left = mid + 1
    else:
        right = mid - 1
    
print(answer)

File: BaekjoonOnlineJudge/2512/2512.java
import java.io.*;
import java.util.*;


public class Main {

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;
    private static int n, m;
    private static int[] budgets;

    private static int getDistributedBudget(int maxiumBudget) {
        int distributedBudget = 0;

        for (int budget : budgets) {
            if (budget > maxiumBudget) {
                distributedBudget += maxiumBudget;
            } else {
                distributedBudget += budget;
            }
        }

        return distributedBudget;
    }

    public static void main(String[] args) throws IOException {
        n = Integer.parseInt(br.readLine());
        budgets = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            budgets[i] = Integer.parseInt(st.nextToken());
        }
        m = Integer.parseInt(br.readLine());

        int answer = 0;
        int left = 1;
        int right = Arrays.stream(budgets).max().getAsInt();

        while (left <= right) {
            int mid = (left + right) / 2;

            if (getDistributedBudget(mid) > m) {
                right = mid - 1;
            } else {
                answer = Math.max(answer, mid);
                left = mid + 1;
            }
        }

        bw.write(String.valueOf(answer));

        bw.flush();
        bw.close();
        br.close();
    }
}